### DISCUSSION TITLE
m1: act fails to pull with unauthorized: incorrect username or password

### DISCUSSION THREAD

**Discussion by ssbarnea** [nektos/act#1165] [General]
I am on MacOS M1 and act failed to even pull the container reporting not being able to login to docker registry, but docker works correctly and it is able to pull.

```
$ act
WARN  ⚠ You are using Apple M1 chip and you have not specified container architecture, you might encounter issues while running act. If so, try running it with '--container-architecture linux/amd64'. ⚠
ERRO[0000] Unable to interpolate expression 'format('{0}', matrix.env.TOXENV)': Unable to dereference 'toxenv' on non-struct 'invalid'
ERRO[0000] 'runs-on' key not defined in push/ack
ERRO[0000] 'runs-on' key not defined in push/ack
[tox/lint-1     ] 🧪  Matrix: map[env:map[TOXENV:lint] os:ubuntu-20.04 python-version:3.9]
[tox/py-3       ] 🧪  Matrix: map[env:map[TOXENV:py] os:ubuntu-20.04 python-version:3.9]
[tox/packaging-2] 🧪  Matrix: map[env:map[TOXENV:packaging] os:ubuntu-20.04 python-version:3.9]
[tox/py-3       ] 🚀  Start image=node:16-buster-slim
[tox/packaging-2] 🚀  Start image=node:16-buster-slim
[tox/lint-1     ] 🚀  Start image=node:16-buster-slim
[tox/py-3       ]   🐳  docker pull image=node:16-buster-slim platform= username= forcePull=false
[tox/packaging-2]   🐳  docker pull image=node:16-buster-slim platform= username= forcePull=false
[tox/lint-1     ]   🐳  docker pull image=node:16-buster-slim platform= username= forcePull=false
Error: Error response from daemon: Head "https://registry-1.docker.io/v2/library/node/manifests/16-buster-slim": unauthorized: incorrect username or password
FAIL: 1

ssbarnea@m1: ~/c/schemastorepy main
$ docker pull node:16-buster-slim
16-buster-slim: Pulling from library/node
383d573fdcf2: Pull complete
a3942ef45b19: Pull complete
233507218b05: Pull complete
d14f918387bf: Pull complete
9f12e5ac2608: Pull complete
Digest: sha256:e5936cf8dbbb3c0d9d337c0e9f23e56e00b7e62280a8a617d12adfdda8ca9f11
Status: Downloaded newer image for node:16-buster-slim
docker.io/library/node:16-buster-slim

ssbarnea@m1: ~/c/schemastorepy main
$ docker --version
Docker version 20.10.14, build a224086
```

**Comment by catthehacker** [comment]
Can you show your workflow and secrets/env?
We use in-workflow credentials, credentials passed into act and docker cli/desktop credential helpers.
https://github.com/nektos/act/blob/91fd412c51667dd10c434e9eb16ba10685581fa6/pkg/runner/run_context.go#L657-L658
https://github.com/nektos/act/blob/91fd412c51667dd10c434e9eb16ba10685581fa6/pkg/runner/run_context.go#L680
https://github.com/nektos/act/blob/b910a42edfab7a02b08a52ecef203fd419725642/pkg/container/docker_auth.go

**Reply by ssbarnea** [reply]
There are no secrets involved, the repository was https://github.com/ansible/ansible-lint so it should be easy to reproduce. I double checked that I do not have any `DOCKER_*` variables defined.

I do suspect that something weird happens on how the docker is called as I observed that the logged line is not really the effective command being executed as it does not have valid syntax.

Not sure if my docker config file is what is confusing act about how it calls docker?
```
$ cat ~/.docker/config.json
{
    "auths": {
        "https://index.docker.io/v1/": {},
        "registry.redhat.io": {
            "auth": "..."
        },
        "trunk.registry.rdoproject.org": {}
    },
    "credsStore": "desktop",
    "experimental": "enabled",
    "stackOrchestrator": "swarm"
}%
```

As you can see the credsStore is enable as that is where the secrets are supposed to be stored.

Basically I was not able to make `docker pull` from cli to fail, but act failed twice in a row. Still, after doing the pulls myself, act proceeded. I concluded that act is unable to make the pulls itself.

**Reply by catthehacker** [reply]
> I do suspect that something weird happens on how the docker is called as I observed that the logged line is not really the effective command being executed as it does not have valid syntax.

We do not run Docker as CLI but as API. Those printed commands are just for reference of what we are doing internally and are extremely simplified.

> I concluded that act is unable to make the pulls itself.

Not entirely sure where is the problem. I can't reproduce this, I assume it's something about credential store.

@ZauberNerd / @KnisterPeter implemented it so maybe they have an idea. Maybe it grabs credentials for different registry instead of `index.docker.io`.

**Reply by ssbarnea** [reply]
I will try to call the API myself tomorrow and see if I get a different result, or maybe they will know more about it.

The issue far from urgent as there is a workaround for it, but we might still want to fix it once we identify the cause.

Thanks!

**Reply by catthehacker** [reply]
ref: https://github.com/nektos/act/issues/1166

**Reply by KnisterPeter** [reply]
I'm able to run act without credentials and without pre-existing container:

This is my workflow. It does pull `node:12-buster-slim` with my `.actrc`.

`.actrc`:

```
-P ubuntu-latest=node:12-buster-slim
-P ubuntu-20.04=node:12-buster-slim
-P ubuntu-18.04=node:12-buster-slim
-P ubuntu-16.04=node:12-stretch-slim
```

`workflow.yml`:

```yaml
name: asdf
on: push
jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - run: echo test
```

@ssbarnea Can you try to trim your example to the bare minimum as well? The test workflow from you as multiple dependencies (e.g. a matrix) which might have a lot of influence on that.

**Reply by nikhilbadyalanchanto** [reply]
Unable to run even this setup.

**Reply by KnisterPeter** [reply]
@nikhilbadyalanchanto Do you use a Mac M1 machine as well?

**Reply by nikhilbadyal** [reply]
@KnisterPeter  Yes, I'm also on MAC M1. Btw I'm the same person from above comment. Just diff account

**Reply by ZauberNerd** [reply]
@nikhilbadyal what does your log say when you run act with `-v`? Any errors/warnings? Or the same as above?

**Reply by nikhilbadyal** [reply]
`test` is the folder name. I added DOTS before that to hide path.

```
DEBU[0000] Loading environment from ....../test/.env
DEBU[0000] Loading secrets from....../test/.secrets
DEBU[0000] Loading workflows from '....../test/.github/workflows'
DEBU[0000] Loading workflows recursively
DEBU[0000] Found workflow 'ci.yml' in '...../test/.github/workflows/ci.yml'
DEBU[0000] Reading workflow '....../test/.github/workflows/ci.yml'
DEBU[0000] Planning event: push
WARN[0000] unable to get git repo: unable to find git repo
WARN[0000] unable to get git ref: unable to find git repo
WARN[0000] unable to get git revision: unable to find git repo
DEBU[0000] evaluating expression 'success()'
DEBU[0000] expression 'success()' evaluated to 'true'
[CI/test] 🚀  Start image=node:12-buster-slim
WARN[0000] unable to get git repo: unable to find git repo
WARN[0000] unable to get git ref: unable to find git repo
WARN[0000] unable to get git revision: unable to find git repo
[CI/test]   🐳  docker pull image=node:12-buster-slim platform=linux/amd64 username= forcePull=false
[CI/test]   🐳  docker pull node:12-buster-slim
DEBU[0000] Image exists? false
[CI/test] pulling image 'docker.io/library/node:12-buster-slim' (linux/amd64)
WARN[0002] unable to get git repo: unable to find git repo
WARN[0002] unable to get git ref: unable to find git repo
WARN[0002] unable to get git revision: unable to find git repo
Error: Error response from daemon: Head "https://registry-1.docker.io/v2/library/node/manifests/12-buster-slim": unauthorized: please use personal access token to login
```

ci.yml
```
name: CI
on: push
jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - run: echo test

```
_actrc
```
-P ubuntu-latest=node:12-buster-slim
-P ubuntu-20.04=node:12-buster-slim
-P ubuntu-18.04=node:12-buster-slim
--container-architecture linux/amd64
```

**Reply by catthehacker** [reply]
Would be better to see Docker config if you have any

**Reply by ZauberNerd** [reply]
@nikhilbadyal and what's the content (**make sure to redact confidential information**) of your `~/.docker/config.json`?

Also, just as an experiment: Could you try to rename/move the docker config and try to run act again? Public docker hub usually should not require authentication, but we are using credentials, if they are configured.
And as written here: https://github.com/nektos/act/discussions/1165#discussioncomment-2757988 it looks like act might select the wrong credentials. So by trying it out without a docker config, we could verify, that the issue is indeed related to the authentication code.

**Reply by nikhilbadyal** [reply]
`cat ~/.docker/config.json`

```
{
	"auths": {
		"https://index.docker.io/v1/": {}
	},
	"credsStore": "desktop"
}
```

**Reply by nikhilbadyal** [reply]
Even after removing `config.json` to `fgerg.json`

```
➜  test act -v
DEBU[0000] Loading environment from ..../test/.env
DEBU[0000] Loading secrets from ..../test/.secrets
DEBU[0000] Loading workflows from '....test/.github/workflows'
DEBU[0000] Loading workflows recursively
DEBU[0000] Found workflow 'ci.yml' in '..../test/.github/workflows/ci.yml'
DEBU[0000] Reading workflow '.../test/.github/workflows/ci.yml'
DEBU[0000] Planning event: push
WARN[0000] unable to get git repo: unable to find git repo
WARN[0000] unable to get git ref: unable to find git repo
WARN[0000] unable to get git revision: unable to find git repo
DEBU[0000] evaluating expression 'success()'
DEBU[0000] expression 'success()' evaluated to 'true'
[CI/test] 🚀  Start image=node:12-buster-slim
WARN[0000] unable to get git repo: unable to find git repo
WARN[0000] unable to get git ref: unable to find git repo
WARN[0000] unable to get git revision: unable to find git repo
[CI/test]   🐳  docker pull image=node:12-buster-slim platform=linux/amd64 username= forcePull=false
[CI/test]   🐳  docker pull node:12-buster-slim
DEBU[0000] Image exists? false
[CI/test] pulling image 'docker.io/library/node:12-buster-slim' (linux/amd64)
WARN[0003] unable to get git repo: unable to find git repo
WARN[0003] unable to get git ref: unable to find git repo
WARN[0003] unable to get git revision: unable to find git repo
Error: Error response from daemon: Head "https://registry-1.docker.io/v2/library/node/manifests/12-buster-slim": unauthorized: please use personal access token to login
```

**Reply by ZauberNerd** [reply]
Hm. That's strange...
If you have golang installed, you could clone this repository: https://github.com/ZauberNerd/docker-pull-test and execute `go run main.go` (remove the `node:12` image before by running `docker rmi node:12-buster-slim`).

**Reply by nikhilbadyal** [reply]
This is working. I was able to clone

**Comment by ZauberNerd** [comment]
Hmm. At a first glance I don't see why this should fail.

It could be that our auth code is a bit too naive: https://github.com/nektos/act/blob/master/pkg/container/docker_auth.go#L29

In the docker CLI there's a comment about legacy hostnames: https://github.com/docker/cli/blob/master/vendor/github.com/docker/docker/registry/config.go#L25-L47 
maybe we need to expand our auth code a bit more.

I'd be interested to know what the values of `hostname` and `authConfig` are here: https://github.com/nektos/act/blob/master/pkg/container/docker_auth.go#L29

**Comment by nikhilbadyalanchanto** [comment]
Facing same issue.

**Comment by ssbarnea** [comment]
Based on on my recent experience, I would say that GHA has zero support for `arm64` architecture. This means that it makes almost no sense not to automatically add `--container-architecture linux/amd64` switch instead of asking act user to add it.

Second issue is that now I am looking at a static line about pulling and I have no idea if the pull is really happening or not as there is zero feedback on the console. Looking a CPU load does not give me much feedback either. Based on Activity Monitor I do see a little bit of incoming traffic from `com.docker.vpnkit` process, so I can guess that probably it is downloading. still, no idea if it will finish before I type this, in an hour or never. CLI docker does display more info, so I am inclined to find the use of API for pulling as a notable downgrade.

I guess `[check/check]   🐳  docker pull image=ghcr.io/catthehacker/ubuntu:full-20.04 platform= username= forcePull=false` image could be big...

I will be back with more info.

**Reply by ZauberNerd** [reply]
Running act with `-v` gives more status about pulling.

**Reply by catthehacker** [reply]
> I guess `[check/check] 🐳 docker pull image=ghcr.io/catthehacker/ubuntu:full-20.04 platform= username= forcePull=false` image could be big...

It's about 40GB, and 20GB when compressed

**Reply by alaugks** [reply]
Thank you for hint!

**Comment by cinder92** [comment]
i found a work around, run following commands

```
docker login --username your_username_here
docker pull node:16-buster-slim
act
```

that worked for me

**Comment by mloskot** [comment]
I've just installed  act 0.2.43 and trying the test run according to the `README.md` using the `github-actions-demo` and it's failing with this:

```
$ cd D:\\github-actions-demo
$ act
[CI/test] 🚀  Start image=catthehacker/ubuntu:act-latest
[CI/test]   🐳  docker pull image=catthehacker/ubuntu:act-latest platform= username= forcePull=true
Error: Error response from daemon: Head "https://registry-1.docker.io/v2/catthehacker/ubuntu/manifests/act-latest": unauthorized: incorrect username or password
```

I'm reading the docs and I can not find any authentication prerequisites for the demo.

Is there any secret required?

**Reply by catthehacker** [reply]
There isn't, but we do not manage the registries that host images so it could be rate-limiting on their side

**Reply by mloskot** [reply]
Right, I didn't thought this could be hitting the limits. Thank you @catthehacker for the answer

**Comment by EmmanuelTramoy** [comment]
Hi, similar issue working with my own workflow, on a Windows 10 machine :

```bash
[Is Class checkbox checked ?/check-checkbox] [DEBUG] evaluating expression 'success()'
[Is Class checkbox checked ?/check-checkbox] [DEBUG] expression 'success()' evaluated to 'true'
[Is Class checkbox checked ?/check-checkbox] 🚀  Start image=catthehacker/ubuntu:act-latest
[Is Class checkbox checked ?/check-checkbox]   🐳  docker pull image=catthehacker/ubuntu:act-latest platform= username= forcePull=true
[Is Class checkbox checked ?/check-checkbox] [DEBUG]   🐳  docker pull catthehacker/ubuntu:act-latest
[Is Class checkbox checked ?/check-checkbox] [DEBUG] pulling image 'docker.io/catthehacker/ubuntu:act-latest' ()
time="2023-03-16T10:26:25+01:00" level=debug msg="Saving notices etag=612941e9-ab25-4fd7-83eb-1d15267c66b1"
time="2023-03-16T10:26:25+01:00" level=debug msg="No new notices"
Error: Error response from daemon: Head "https://registry-1.docker.io/v2/catthehacker/ubuntu/manifests/act-latest": unauthorized: incorrect username or password
```
I can login with `docker login` and pull the docker image (says it's up to date). Tried to refresh the credentials in credential manager, no changes.

**Comment by ChristopherHX** [comment]
I saw this problem has no solution for a long time, while I never had problems myself.

Can you try this change https://github.com/nektos/act/pull/1774?

It logs now if it uses DockerAuthConfig credentials and if it fails pulling while authenticated it removes the credentials and tries again. However this still means you should remove stale docker auth data whereever it is located.

**Reply by lion24** [reply]
Some issue as other from a WSL2 ubuntu distro on a Windows 11 machine.

I confirm this is working fine with your patch in my setup.

Kr.

**Comment by thirtytwobits** [comment]
I'm seeing this as well on a mac (intel).

**Comment by sombriks** [comment]
removing `.docker/config.json` solved my issue, i had docker-desktop installed on this machine for a while but i removed that because it was using a different context. the file was dangling here afer i removed it.

**Comment by jekru** [comment]
Remove `credsStore` from **config.json**, then use `docker login`. Instead of `docker login`, you can also manually add this to the config:

```json
"auths": {
   "https://index.docker.io/v1/": {
      "auth": ""
   }
}
```

The value of `auth` equals "`username:password`" but encoded in base64.

**Comment by volkmarnissen** [comment]
Uninstall and reinstall of docker using brew install --cask docker solved it for me..

All docker pull commands were failing with the same message.
